src/rocketmq_producer.erl

```diff
@@ -85,50 +88,86 @@ batch_send_sync(Pid, Messages, Timeout) ->
     case gen_tcp:connect(Host, Port, merge_opts(Opts, ?TCPOPTIONS), ?TIMEOUT) of
         {ok, Sock} ->
             tune_buffer(Sock),
             gen_tcp:controlling_process(Sock, self()),
             start_keepalive(),
-            {next_state, connected, State#state{sock = Sock}};
+            ClientID = client_id(Sock),
+            case maybe_get_tls_sock(Sock, SSLOpts) of
+                {error, SSLError} ->
+                    {stop, {shutdown, SSLError}, State};
+                Sock1 ->
+                    {next_state, connected, State#state{sock = Sock1, client_id = ClientID}}
+            end;
         Error ->
             {stop, {shutdown, Error}, State}
     end;
 
 idle(_, ping, State = #state{sock = undefined}) ->
     {keep_state, State}.
 
+maybe_get_tls_sock(Sock, undefined) ->
+    Sock;
+maybe_get_tls_sock(Sock, SSLOpts) ->
+    case ssl:connect(Sock, SSLOpts, ?TIMEOUT) of
+        {ok, Sock1} ->
+            Sock1;
+        Error ->
+            Error
+    end.
+
 connected(_EventType, {tcp_closed, Sock}, State = #state{sock = Sock}) ->
     log_error("TcpClosed producer: ~p~n", [self()]),
     erlang:send_after(5000, self(), connecting),
     {next_state, idle, State#state{sock = undefined}};
 
+connected(_EventType, {ssl_closed, Sock}, State = #state{sock = Sock}) ->
+    log_error("SSLClosed producer: ~p~n", [self()]),
+    erlang:send_after(5000, self(), connecting),
+    {next_state, idle, State#state{sock = undefined}};
+
+connected(_EventType, {ssl_error, Sock, Reason}, State = #state{sock = Sock}) ->
+    log_error("SSL Socket Error, producer: ~p, reason: ~p~n", [self(), Reason]),
+    erlang:send_after(5000, self(), connecting),
+    {next_state, idle, State#state{sock = undefined}};
+
 connected(_EventType, {tcp, _, Bin}, State) ->
     handle_response(Bin, State);
 
+connected(_EventType, {ssl, _, Bin}, State) ->
+    handle_response(Bin, State);
+
 connected({call, From}, {send, MsgAndProps}, State = #state{sock = Sock,
                                                         topic = Topic,
                                                         queue_id = QueueId,
                                                         producer_group = ProducerGroup,
                                                         opaque_id = Opaque,
                                                         requests = Reqs,
-                                                        producer_opts = ProducerOpts
+                                                        producer_opts = ProducerOpts,
+                                                        send_sock_mod = SendSockMod
                                                         }) ->
-    send(Sock, ProducerGroup, get_namespace(ProducerOpts), Topic, Opaque, QueueId, MsgAndProps, get_acl_info(ProducerOpts)),
+    send(Sock, ProducerGroup, get_namespace(ProducerOpts), Topic, Opaque, QueueId, MsgAndProps, get_acl_info(ProducerOpts), SendSockMod),
```

Comment:
Shall we check the ret val? or just do `ok=send( ...`

Reply:
Not sure. I did not write the original code so I don't understand it 100% but I will check what the return value of send could be.

Reply:
send can return an error tuple. This is currently just ignored. This existed before this PR so its not a new problem but its probably not a good behavior since errors might not be noticed. Will check our connector code as well.

Reply:
Fixed by making sure that the response from SocketMod:send call is returned to the caller.